### PR TITLE
[SPARK-14993] [SQL] Fix Partition Discovery Inconsistency when Input is a Path to Parquet File 

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

When we load a dataset, if we set the path to `/path/a=1`, we will not take `a` as the partitioning column. However, if we set the path to `/path/a=1/file.parquet`, we take `a` as the partitioning column and it shows up in the schema. 

This PR is to fix the behavior inconsistency issue.

The base path contains a set of paths that are considered as the base dirs of the input datasets. The partitioning discovery logic will make sure it will stop when it reaches any base path.

By default, the paths of the dataset provided by users will be base paths. Below are three typical cases,
**Case 1**`sqlContext.read.parquet("/path/something=true/")`: the base path will be
`/path/something=true/`, and the returned DataFrame will not contain a column of `something`.
**Case 2**`sqlContext.read.parquet("/path/something=true/a.parquet")`: the base path will be
still `/path/something=true/`, and the returned DataFrame will also not contain a column of
`something`.
**Case 3**`sqlContext.read.parquet("/path/")`: the base path will be `/path/`, and the returned
DataFrame will have the column of `something`.

Users also can override the basePath by setting `basePath` in the options to pass the new base
path to the data source. For example, 
`sqlContext.read.option("basePath", "/path/").parquet("/path/something=true/")`,
and the returned DataFrame will have the column of `something`.

The related PRs: 
- https://github.com/apache/spark/pull/9651
- https://github.com/apache/spark/pull/10211
#### How was this patch tested?

Added a couple of test cases

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala [hunk 1]
```diff
@@ -184,8 +184,10 @@ private[sql] object PartitioningUtils {
         return (None, None)
       }
 
-      if (basePaths.contains(currentPath)) {
+      if (basePaths.contains(currentPath) ||
+        basePaths.exists(_.toString.startsWith(currentPath.toString))) {
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Can you explain this and provide an example?

**Comment by yhuai** [comment]
@gatorsmile When we call PartitioningUtils.parsePartitions, we should provide a `Seq[Path]` representing leaf dirs, right? We have this problem is caused by the fact we actually pass leaf files in?

**Comment by gatorsmile** [comment]
@yhuai We passed leaf dirs to `path`, but the `basePaths` is a path to a `Parquet` file. For example, 
```
    parsePartition(
      path = new Path("file://path/a=10"),
      defaultPartitionName = defaultPartitionName,
      typeInference = true,
      basePaths = Set(new Path("file://path/a=10/p.parquet")))
```

In this case, we need to follow what we did in https://github.com/apache/spark/pull/9651. 

The current behavior is shown in the test case:
https://github.com/apache/spark/pull/12828/files#diff-cf57fe1c329fb21ac00a8528f049da4aR435

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala (hunk 1 above)
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: I see. We are trying to check if there is a basePath starts with the currentPath. 

So, the actual problem is that `basePaths` in HDFSFileCatalog contains files, right? I discussed it with @tdas. He will have a pr to change `basePaths`. Let's review his fix together. What do you think?

Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/PartitioningUtils.scala]: Sure, please include the test case in 
https://github.com/apache/spark/pull/12828/files#diff-cf57fe1c329fb21ac00a8528f049da4aR435

This test case checks three typical cases.

**Comment by tdas** [comment]
@yhuai and I discussed that this solution of substring match seems very hacky. 

The real problem is that basePaths should never have files as it does not make sense to have a basePath that is not a directory. So, our strategy in HDFSFileCatalog of making the set of input files as the default basePath is incorrect. The correct fix is to set the default base path based on the [dirs in input paths] UNION [parent dirs of files in input paths]. 

Here is the fix - https://github.com/apache/spark/commit/fbef90f47db7c0a81ec29db27e83d0daf56673bd
Please update your PR with this. You dont have to change `parsePartition` in that case. 

Consider updating the scala docs to make this implicit assumption of `basePath` clear in the code.

**Comment by gatorsmile** [comment]
@tdas Thank you very much! Will do it soon.

**Comment by gatorsmile** [comment]
@tdas @yhuai Based on the fix https://github.com/apache/spark/commit/fbef90f47db7c0a81ec29db27e83d0daf56673bd , updated the scala docs, test cases and PR description.

Please let me know if anything is not appropriate. Thanks again!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala [hunk 2]
```diff
@@ -423,23 +423,34 @@ class HDFSFileCatalog(
...
+    parameters.get("basePath").map(new Path(_)) match {
+      case Some(userDefinedBasePath) =>
+        val fs = userDefinedBasePath.getFileSystem(hadoopConf)
+        if (!fs.isDirectory(userDefinedBasePath)) {
+          throw new IllegalArgumentException("Option 'basePath' must be a directory")
+        }
+        Set(userDefinedBasePath.makeQualified(fs.getUri, fs.getWorkingDirectory))
+
+      case None =>
+        paths.map { path => if (leafFiles.contains(path)) path.getParent else path }.toSet
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: Do we need to make this `path` qualified?

Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: `leafFiles` only contain qualified paths, right?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala [hunk 3]
```diff
@@ -423,23 +423,34 @@ class HDFSFileCatalog(
...
    */
   private def basePaths: Set[Path] = {
-    val userDefinedBasePath = parameters.get("basePath").map(basePath => Set(new Path(basePath)))
-    userDefinedBasePath.getOrElse {
-      // If the user does not provide basePath, we will just use paths.
-      paths.toSet
-    }.map { hdfsPath =>
-      // Make the path qualified (consistent with listLeafFiles and listLeafFilesInParallel).
-      val fs = hdfsPath.getFileSystem(hadoopConf)
-      hdfsPath.makeQualified(fs.getUri, fs.getWorkingDirectory)
```
Comment (yhuai) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: I am wondering why I did not just call `makeQualified(fs)` when I wrote this part...

Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: I was wondering the same multiple times in other PRs.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala (hunk 2 above)
Comment (tdas) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: I believe leaf files contain only qualified. There was comments elsewhere in the file that same so. 

Here it is - https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala#L468

**Comment by tdas** [comment]
just a heads up ... i have a PR that refactors FileCatalog significantly - #12879 . I want to merge that first which is cause conflicts in this PR as well as my PR #12856

**Comment by gatorsmile** [comment]
Sure, I will wait for it. Thanks for letting me know it!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: Will change it to `fs.makeQualified(userDefinedBasePath)`. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/fileSourceInterfaces.scala]: Will make `path` qualified before comparison. Thanks!

**Comment by yhuai** [comment]
LGTM. Merging to master and branch 2.0

**Pull request closed without merging**